### PR TITLE
Group 1 Extensible Calculator

### PR DESCRIPTION
**Pull request by mlech456**
- Aron Ivarsson
- Kekoa Iobst
- Mark Lechmanik

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -60,3 +60,6 @@ target/
 
 #Ipython Notebook
 .ipynb_checkpoints
+
+#calculator
+calculator/
```
Comment (santiagobasulto) [inline .gitignore:65]: 🤔 You guys want to ignore the project dir?

File: calculator/exceptions.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 
 class InvalidOperation(Exception):
-    pass
+   """Raises an error when an invalid operation is passed to a calculator """
```
Comment (santiagobasulto) [inline calculator/exceptions.py:3]: It's a good practice to also include `pass` even if you have documentation.

File: calculator/main.py [hunk 3]
```diff
@@ -26,8 +33,17 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
-
+    for i in params:
+        if not isinstance(i, (int,float)):
+            raise InvalidParams("Given params are invalid.")
```
Comment (santiagobasulto) [inline calculator/main.py:38]: Good! Can you guys think a more Pythonic solution? Maybe with a list comprehension?

File: calculator/main.py [hunk 4]
```diff
@@ -26,8 +33,17 @@ def perform_operation(calc, operation, params):
...
+    
+    total = calc['operations'][operation](*params)
+
+    calc['history'].append((str(datetime.now()),operation,params,total))
```
Comment (santiagobasulto) [inline calculator/main.py:42]: Separate the arguments with an space. It's just a formatting/style from PEP8: https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

``` python
calc['history'].append((str(datetime.now()), operation, params, total))
```

File: calculator/main.py [hunk 5]
```diff
@@ -37,14 +53,18 @@ def add_new_operation(calc, operation):
     :param operation: Dict with the single operation to be added.
                       ie: {'add': add_function}
     """
-    pass
+    if not isinstance(operation, dict):
+        raise InvalidOperation("Given operation is invalid.")
```
Comment (santiagobasulto) [inline calculator/main.py:57]: This code looks familiar (from `create_new_calculator`)... Can you maybe avoid repetition?

File: calculator/main.py [hunk 6]
```diff
@@ -37,14 +53,18 @@ def add_new_operation(calc, operation):
...
+    return calc['operations'].update(operation)
 
 
 def get_operations(calc):
     """
     Returns the list of operation names supported by given calculator.
     """
-    pass
+    
+    return list(calc['operations'].keys())
```
Comment (santiagobasulto) [inline calculator/main.py:67]: Nice, fixing Python 3 😉

File: calculator/main.py [hunk 7]
```diff
@@ -58,18 +78,24 @@ def get_history(calc):
...
 
 def repeat_last_operation(calc):
     """
     Returns the result of the last operation executed in the history.
     """
-    pass
+    if calc['history'] != []:
+        return calc['history'][-1][3]               
+    else:
+        return None
```
Comment (santiagobasulto) [inline calculator/main.py:99]: Can you guys think in a one line solution for this? (just for learning purposes, so you smash your head against the wall for a few minutes)

**Comment by santiagobasulto** [comment]
Really nice implementation!

File: calculator/exceptions.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 
 class InvalidOperation(Exception):
```
Comment (prestonr83) [inline calculator/exceptions.py:2]: Wasn't aware you could use a empty class for an Exception.

File: calculator/operations.py [hunk 9]
```diff
@@ -1,20 +1,29 @@
...
 
 def subtract(*args):
     # your implementation here
-    pass
-
+    x = float(args[0])
+    for arg in args[1:]:
+        x = x-float(arg)
+    return x
+
```
Comment (prestonr83) [inline calculator/operations.py:12]: Could use reduce functions on most of these for less lines but you'd also have to account for using Python 3.

**Pull request closed without merging**